docs/proposals/registry/devfile-packaging.md

```diff
@@ -0,0 +1,51 @@
+
+<img width="742" alt="Screen Shot 2020-10-22 at 1 54 00 PM" src="https://user-images.githubusercontent.com/6880023/97219676-b2637200-17a0-11eb-8465-1063aa048768.png">
+
+
+Each folder under devfiles corresponds to a stack, with its own devfile and associated artifacts. There may be an index.json present for devfile consumers
+
+From this point, the Github repository may act as self-hosting devfile registry (as in the case of [odo-devfiles/registry](https://github.com/odo-devfiles/registry), or it may be built up into an Apache server container (as in the case of v1 registries like [eclipse/che-devfile-registry](https://github.com/eclipse/che-devfile-registry)).
+
+## Are any changes needed to accommodate OCI?
+Storing the devfiles (and any associated artifacts) on a dedicated Github repository (the “devfile registry repository”), is still a good idea.
+
+   - Provides a clear, straightforward and non-cluttered place for stack owners to contribute devfiles to
+   - Separates the devfile stacks from the OCI registry code and logic
+
+But, with the move to OCI registries, we need a way to easily package the devfile stacks from GitHub and load them into the OCI registry
+
+   - Needs to retrieve the devfiles without an internet connection
+     - Means we can’t have the metadata container git clone the devfiles before pushing them up
+   - Needs to be easily built and distributable by registry admins.  
+     - Once we support “deploying your own registry”, with custom devfiles this will become especially important
+   - Shouldn’t require building and deploying custom components of the registry to distribute (i.e. it shouldn’t be baked into the registry or metadata container images)
+     - Currently, the metadata container has the devfiles built into it, meaning users need to build their own metadata container image to customize devfiles
+
+## Proposed Changes
+
+**To address the requirements listed above, rather than bundling the devfiles as part of registry's bootstrap container, build the “devfile registry repository” into its own docker image**
+
+<img width="636" alt="Screen Shot 2020-10-26 at 3 39 33 PM" src="https://user-images.githubusercontent.com/6880023/97220180-767cdc80-17a1-11eb-85fd-cf1a5a623aeb.png">
+
+
+This container image will contain just the devfiles (and their artifacts), as well as any other required files.
+
+When deploying the OCI registry, the image will be specified as an init container, with an entrypoint command that copies the devfiles (and artifacts) to the bootstrap container’s volume. Then, as-is today, when the bootstrap container starts up, it generates the index.json and pushes the devfiles up into the OCI registry using `oras`.
```

Comment:
I don't know if this is underspecified or sufficient for now, but do we need to describe some sort of contract between the init container and the registry itself? It might help to have clarity in how the init command "glue" is going to be defined, e.g. with "env var `X` defines where devfiles are stored in the init container` or `env var `X` defines the path to devfiles in the registry`. 

Can I plop multiple init containers onto the registry and have them merge into one? This would be useful for e.g. adding a few of my own devfiles while keeping the default as-is. Having a defined description of where plugins should be and how they will be copied means potentially less work for init container authors.

Reply:
That's a good question. I could probably add the following to the proposal to make things more clear:

> The following environment variables will be set in the init container image to define where the devfiles and plugins are to be copied from:
`DEVFILES_SOURCE=/devfiles`
`PLUGINS_SOURCE=/plugins`

and 

> The follow entrypoint command will be used to copy the devfiles and plugin over to the registry's volume:
`cp -rf $DEVFILES_SOURCE /registry && cp -rf $PLUGINS_SOURCE /registry`


Re multiple init containers: I hadn't thought of that, but I see no reason why we _couldn't_ allow it, and agreed that it could be useful as an alternative to modifying the default image. Maybe have something like the following in the registry custom resource?

```
devfileSource:
    - "quay.io/devfile/devfiles:latest"
    - "docker.io/johncollier/somedevfiles:latest"
    - "docker.io/someorg/moredevfiles:latest"
```

And when the registry is deployed, each of those entries map to an init container, copying the contents into the registry's volume?

Reply:
So Elson and I discussed things today and we decided to not go forward with the initContainer model, but rather package the devfiles into the bootstrap image in the registry